app/models/device_token.rb

```diff
@@ -0,0 +1,16 @@
+class DeviceToken < ActiveRecord::Base
+  attr_accessible :token
+
+  def save
+    # Escape token: remove arrows and spaces
+    escaped_token = ""
+    token.each_char do |c|
+      escaped_token += c if c != " " && c != "<" && c != ">"
+    end
+    self.token = escaped_token
+
+    # Save if it does not already exist
+    return if DeviceToken.where("token = ?", self.token).count > 0
```

Comment:
In other words, you need to return gracefully when the constraint is violated, that is no problem. But as it stands there is a race condition here if two clients send the same token at the same time (probably not a big deal with our traffic ;))